lib/pbench/server/database/models/datasets.py

```diff
@@ -687,61 +687,78 @@ class Metadata(Database.Base):
 
     __tablename__ = "dataset_metadata"
 
+    # "Native" keys are the value of the PostgreSQL "key" column in the SQL
+    # table. We support hierarchical nested keys of the form "server.indexed",
+    # but the first element of any nested key path must be one of these:
+    NATIVE_KEYS = ["dashboard", "server", "user"]
```

Comment:
> with separate namespaces the dashboard UI code doesn't need to worry about restricting the potential user key namespace when we add a UI for "labels" or "keywords" or whatever we call them.

From the dashboard perspective, I think it's six of one and a half dozen of the other:  the "label" goes under either `user.labels` or under `dashboard.user_labels` (or just `dashboard.labels`).  That is, it would be (I think...) just as easy for the dashboard code to put the contents in a subtree of the `dashboard` space as it would be to put it in the `user` space.

But, from the server perspective, or some other global perspective, there's a big difference.  We really should be making it easy for the server to support multiple clients with a minimum of effort.  Having a client-specific namespace for each client (i.e., `dashboard` for the Pbench Dashboard) doesn't seem unreasonable.  However, having a _common_ namespace (`user`) seems like a disaster waiting to happen, unless we impose some sort of architecture on it.  Rather than architecting that space, we could instead just let each client put what they want in their own namespaces and not bother providing the `user` space.

> It seemed like a natural and useful segregation.

It is from the Dashboard perspective, but not from the Server perspective.  And, that segregation should be easy enough (for the Dashboard to create on its own with suitable support from the Server) inside the client-specific namespace.

> I don't really want to over-over-engineer this too much.

I'm sympathetic to the over-engineering concern.  However, the fact that you've already implemented support for the `user` namespace means that it should be very straightforward to support an open `dashboard` namespace instead and just remove the `user` namespace...no additional engineering required!

>  A direct client other than the dashboard can always aggregate its own keys under an ad hoc `user.me` or `user.you` namespace. (Or petition for a new namespace.)

I would suggest that they petition for a new namespace (which, hopefully, is easy to add at this point).  The problem with the alternative is that no one can count on `user.me` or `user.you` being available or being usable without conflict, in the absence of some sort of architecture for the space...which puts us back into the realm of a new client having to submit a petition for space.

> I've tried to build a fairly but not ultimately flexible mechanism here

I don't claim to understand the open namespace implementation, yet, but based on what I've seen, I think you've failed:  I think you have built an ultimately flexible mechanism, here.  😁  I just want to see it put into use cleanly.

Reply:
> From the dashboard perspective, I think it's six of one and a half dozen of the other: the "label" goes under either user.labels or under dashboard.user_labels (or just dashboard.labels). That is, it would be (I think...) just as easy for the dashboard code to put the contents in a subtree of the dashboard space as it would be to put it in the user space.

From my perspective, the main difference is that the server has a stake in the `dashboard` namespace since `PUT` initializes values there; it's not a completely "open" namespace since we're absorbing part of that namespace with a specific semantic. The `user` namespace is completely unrestricted.

Whether I allow users to create first level, second level, or third level keys really has no impact on anything since we can't exercise any real control. And, yes, making `dashboard.*` an open namespace would functionally eliminate the need for a separate `user.*` namespace (or vice versa); but, again, the contract between dashboard and server for initializing `saved` and `seen` on upload reserves part of the namespace. I could have `dashboard.seen` and `dashboard.user.*` but I'm a bit happier to keep this at a separate native key level, and I don't see any reason *not to*.

And if some other API client asks for a new `client.*` namespace (or specific `client.key1` and `client.key2`), it's easy to add with no impact to existing code, independent of how I structure `dashboard` and `user` now.

Reply:
> the contract between dashboard and server for initializing `saved` and `seen` on upload reserves part of the namespace.

Ick.  Any chance we can renegotiate that contract??

For instance, if the Dashboard asks for `saved` and/or `seen` and the server reports that they are undefined, the Dashboard could treat those as false values (or apply whatever defaults it wishes).  If we can all agree on what the initial values _should_ be, then I don't think the Server needs to explicitly set them...and that would release the Server from having to dabble in the `dashboard` space uninvited.  (And, that would allow us to drop the `user` space in favor of the `dashboard` space.)

Reply:
I think having a config setting for the (non-`server`) first-level namespace items is an _excellent_ idea (no, it's a **superb** idea!).

> As for `dashboard`, if you really want to pursue the argument I could be persuaded to convert it to an open namespace reserved (by convention) for the dashboard as a particular client and let it assign and initialize the individual keys as it wants while `user` would be an open playground.

That sounds good.

Reply:
I'm definitely saving the idea of a config setting for later, and I'm still mulling the idea of opening `dashboard` for now. (Whether it becomes just a default config setting later is something I'll deal with when "later" tweaks me on the nose.)